src/azul/service/storage_service.py

```diff
@@ -82,9 +82,13 @@ def upload(self,
         self.client.upload_file(Filename=file_path,
                                 Bucket=self.bucket_name,
                                 Key=object_key,
-                                ExtraArgs=self._object_creation_kwargs(content_type, tagging))
+                                ExtraArgs=self._object_creation_kwargs(content_type))
+        # upload_file doesn't support tags so we need to make a separate request
+        # https://stackoverflow.com/a/56351011/7830612
+        if tagging:
+            self.put_object_tagging(object_key, tagging)
 
-    def _object_creation_kwargs(self, content_type, tagging):
+    def _object_creation_kwargs(self, content_type=None, tagging=None):
```

Comment:
I think this would benefit from type annotations for the parameters.

Reply:
And I suspect that the parameters ought to be made keyword-only now that they're optional, but I'm not certain.

Reply:
I made both changes. I figure it can't hurt!